stablehlo/reference/Ops.cpp

```diff
@@ -30,13 +33,106 @@ namespace mlir {
 namespace stablehlo {
 namespace {
 
+static Type getF64Type(MLIRContext *context) {
+  return mlir::Builder(context).getF64Type();
+}
+
+static Type getI64Type(MLIRContext *context) {
+  return mlir::Builder(context).getI64Type();
+}
+
 Index evalIndices(ArrayRef<Tensor> runtimeIndices) {
   Index index(runtimeIndices.size());
   for (size_t i = 0; i < runtimeIndices.size(); ++i)
     index[i] = runtimeIndices[i].get({}).getIntegerValue().getSExtValue();
   return index;
 }
 
+template <typename T>
+DenseIntElementsAttr getDenseIntElementsAttr(
+    Type elementType, T values,
+    std::optional<SmallVector<int64_t>> valuesShape) {
+  SmallVector<int64_t> shape =
+      valuesShape.has_value()
+          ? *valuesShape
+          : SmallVector<int64_t>({static_cast<int64_t>(values.size())});
+  return DenseIntElementsAttr::get(RankedTensorType::get(shape, elementType),
+                                   values);
+}
+
+static Tensor computeSum(const Tensor &operand, const Tensor &initValue,
+                         const Axes &dimensions, ShapedType resultType) {
+  Tensor result(resultType);
+  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
+       ++resultIt)
+    result.set(*resultIt, initValue.get({}));
+
+  for (auto operandIt = operand.index_begin(); operandIt != operand.index_end();
+       ++operandIt) {
+    Index resultIndex = *operandIt;
+    for (auto dim : llvm::reverse(dimensions))
+      resultIndex.erase(resultIndex.begin() + dim);
+
+    SmallVector<Tensor> args;
+    auto currVal = Tensor(initValue.getType());
+    currVal.set({}, result.get(resultIndex));
+    args.push_back(currVal);
+
+    auto nextVal = Tensor(initValue.getType());
+    nextVal.set({}, operand.get(*operandIt));
+    args.push_back(nextVal);
+
+    auto addResult = evalAddOp(currVal, nextVal, currVal.getType());
+    result.set(resultIndex, addResult.get({}));
+  }
+  return result;
+}
+
+Tensor computeMean(const Tensor &operand, const Axis featureIndex) {
+  auto i64Type = getI64Type(operand.getType().getContext());
+  auto f64Type = getF64Type(operand.getType().getContext());
+
+  SmallVector<int64_t> dimensions(operand.getRank());
+  std::iota(dimensions.begin(), dimensions.end(), 0);
+  dimensions.erase(dimensions.begin() + featureIndex);
+
+  SmallVector<ShapedTypeComponents> inferredReduceType;
+  auto reduceStatus = hlo::inferReduceOp(
+      {}, {operand.getType()}, {RankedTensorType::get({}, f64Type)},
+      getDenseIntElementsAttr(i64Type, dimensions, {}), inferredReduceType);
+
+  if (failed(reduceStatus))
+    report_fatal_error(
+        invalidArgument("Could not infer ReduceOp's return type"));
+
+  auto reducedSum = computeSum(
+      operand,
+      Tensor(RankedTensorType::get({}, f64Type), Element(f64Type, 0.0)),
```

Comment:
Also, the spec needs to be updated to properly express this. Instead of `0.0`, we'll probably want to say `constant(0.0, type(operand))`, and instead of `constant(...)` a few lines below, we'll probably want to say `constant(..., type(operand))`?